File: my-folder/3442-maximum-total-reward-using-operations-i/solution.py
from typing import List

class Solution:
    def maxTotalReward(self, reward: List[int]) -> int:
        n = len(reward)
        reward = sorted(reward)

        dp = [[-1]*4001 for _ in range(n)]
    
        def solve(current, index):
            if index >= n:
                return current
            if dp[index][current]!=-1:
                return dp[index][current]
                
            not_take = solve(current, index + 1)
            take = 0
            if reward[index]>current:
                take =  solve(current+reward[index], index + 1)
            
            ans = max(take, not_take)
            dp[index][current]  = ans
            return dp[index][current]
        
        return solve(0, 0)  

File: my-folder/3442-maximum-total-reward-using-operations-i/solution.cpp
class Solution {
public:
    int solve(vector<int>& nums, int sum, int i,vector<vector<int>> &dp) {
        if (i >= nums.size()) {
            return sum;
        }
        if (dp[i][sum] != -1) {
            return dp[i][sum];
        }
        int ex = solve(nums, sum, i + 1,dp);
        int in = 0;
        if (nums[i] > sum) {
            in = solve(nums, sum + nums[i], i + 1,dp);
        }
        return dp[i][sum] = max(in, ex);
    }

    int maxTotalReward(vector<int>& nums) {
        sort(nums.begin(), nums.end());
          vector<vector<int>>dp(nums.size(),vector<int>(4001,-1));
        return solve(nums, 0, 0,dp);
    }
};
